### PR TITLE
test: Add comprehensive unit tests for speed_sensor.c

### LINKED ISSUE #242: Implement Unit Tests for Speed Calculation Logic
## Summary / Description
Write comprehensive unit tests for the core `calculate_speed_kph()` function. This story focuses on the logic verification: ensuring the formula handles standard inputs, zero speed timeouts, and boundary conditions (overflow/underflow) correctly, leveraging the framework set up in US 1.

## Type / Category
- [x] task
- [ ] bug
- [ ] docs
- [ ] infra
- [ ] spike / research

### Blocked by
- [ ] #241 

### Sub-issues
- [x] #254 
- [x] #256 
- [x] #257 

## Tasks
- [x] Refactor `calculate_speed_kph` to be testable (decouple from direct register access if needed).
- [x] Define test cases for 5 common speeds (e.g., 10, 50, 100 km/h) and assert correct return values.
- [x] Define test cases for zero speed detection (simulating timer timeout).
- [x] Define test cases for boundary conditions (Max RPM, Min RPM, Timer Overflow).
- [x] Execute tests and fix any bugs discovered in the logic.

## Acceptance Criteria (AC)
- [ ] 100% line coverage is achieved for the speed calculation function.
- [x] All implemented unit tests pass successfully.
- [x] Tests cover at least: Normal Operation, Zero/Stop, and Overflow.

## Definition of Done (DoD)
- [x] Tests are written, reviewed, and pass within the CI environment.
- [x] Code refactoring (if any) is reviewed.

## How to validate / Test steps
1.  Run the test suite.
2.  Check the output logs to ensure specific cases (e.g., "Test_Speed_100kph") are listed as PASSED.

### LINKED ISSUE #254: Standard Speed Test Cases
## 1 - Short description
Develop the "Happy Path" test scenarios to verify the basic mathematical correctness of the speed calculation formula (`Speed = Circumference / Period * Scale`). This ensures that for a standard range of input frequencies (representing normal driving speeds), the function returns the expected speed values within an acceptable margin of error.

## 2 - Tasks
- [x] Calculate the expected timer period (ticks) for specific target speeds: 10 km/h, 50 km/h, and 100 km/h based on the wheel circumference constant.
- [x] Write a test case `Test_Speed_10kmh` passing the calculated ticks and asserting the result is 10.
- [x] Write a test case `Test_Speed_50kmh`.
- [x] Write a test case `Test_Speed_100kmh`.
- [x] Run the tests and verify they pass.

## 3 - Acceptance Criteria
- [x] The function accurately calculates speed for all test inputs.
- [x] The calculation error is within the allowed tolerance (e.g., +/- 1 km/h).

## 4 - Notes / Implementation details
Use a tolerance assertion (e.g., `TEST_ASSERT_INT_WITHIN` in Unity or `EXPECT_NEAR` in GoogleTest) to handle floating-point rounding differences if applicable.

### LINKED ISSUE #256: Zero Speed & Timeout Test Cases
## 1 - Short description
Verify the logic handles the "Stationary" condition correctly. Since a stopped car produces infinite period (no pulses), the firmware relies on a "Timeout" or "Max Period" threshold to decide when to force the speed to 0. This test ensures that logic triggers correctly.

## 2 - Tasks
- [ ] Identify the defined `MAX_PERIOD_TICKS` or timeout constant in the code.
- [ ] Write a test case `Test_Speed_Zero_Timeout` that passes a period value slightly larger than the max threshold.
- [ ] Assert that the returned speed is exactly `0`.
- [ ] Verify that no "division by zero" errors occur during this logic path.

## 3 - Acceptance Criteria
- [ ] Inputting a period > Max Threshold results in a return value of 0.
- [ ] The system transitions to 0 km/h without glitching (e.g., showing a massive speed spike before dropping).

## 4 - Notes / Implementation details
If the logic relies on a hardware flag for timeout, you may need to mock that flag or pass it as an argument to the calculation function.

### LINKED ISSUE #257: Boundary & Overflow Test Cases
## 1 - Short description
Boundary & Overflow Test Cases
Stress-test the calculation logic against edge cases to prevent runtime crashes. This includes protecting against Division by Zero (if period is 0), handling 32-bit integer overflows if the period is extremely large, and handling Timer Wraparound (if the math handles raw counter values).

## 2 - Tasks
- [ ] **Div/0 Check:** Write `Test_Input_Zero` passing 0 as the period. Assert the function returns 0 or an error code (does not crash).
- [ ] **Max Speed:** Write `Test_Max_Physical_Speed` passing the smallest possible non-zero period (1 tick). Verify the result doesn't overflow the return variable type.
- [ ] **Timer Wrap:** (If applicable) Write `Test_Timer_Overflow` where `Current_Capture < Last_Capture`, ensuring the math handles the unsigned subtraction correctly.

## 3 - Acceptance Criteria
- [ ] Calling the function with `0` does not cause a Hard Fault or Crash.
- [ ] Extreme inputs return mathematically consistent (or safely clamped) results.

## 4 - Notes / Implementation details
This is critical for safety. In C, `x / 0` is undefined behavior and often crashes the MCU. The code *must* have an `if (period == 0) return ...` check.

### PR DESCRIPTION
**Pull request by berestv**
# Pull Request

**Related issue(s):** closes #242, closes #254, closes #256, closes #257

## Type of change
- [x] test: Adding or updating tests

## Summary
This PR implements comprehensive unit testing for the speed_sensor.c module using Ceedling (Unity + CMock + GCov) framework. The implementation includes:

- **11 test cases** covering different aspects of speed calculation logic:
  - Standard speed calculations
  - Zero speed and timeout scenarios
  - Boundary conditions and overflow handling
  - Edge cases for sensor readings
- **Automated test execution script** (`run_speedtest.sh`) for easy testing and cleanup
- **Comprehensive README.md** documenting the testing strategy, setup, and usage
- **Code coverage reporting** using GCov and LCOV for visualization
- **Mock implementations** for STM32 HAL and ThreadX dependencies to enable host-based testing

The testing framework allows running unit tests on a host PC without requiring the actual STM32 hardware by mocking hardware dependencies.

## How to test / Validation
To run the tests and view results:

```bash
cd tests/unit/speed_sensor
./run_speedtest.sh
```

Or manually with Ceedling:
```bash
cd tests/unit/speed_sensor
ceedling test:all
```

Expected output:
- All 11 tests should pass
- Coverage report is generated in `coverage_report/` directory
- Test results displayed in terminal

## Checklist
- [x] I have run the project tests locally and they pass
- [x] CI checks are green for this branch (or I will fix any failures)
- [x] I have added or updated tests where applicable
- [x] I have updated documentation (if applicable)
- [ ] I have added/updated any migration notes (if database or protocol changes)
- [ ] I have requested appropriate reviewers and added labels if needed
- [x] This PR contains no secrets or sensitive data

## Risks and backward compatibility
None. This PR only adds unit tests and documentation without modifying any production code. The tests are isolated in the `tests/unit/speed_sensor` directory.

## Related / dependent PRs
None.

 ---
**Approval:** Requires a minimum of **2** approvals. <br>
**Action:** The feature branch **MUST** be deleted upon successful merge.

### CODE REVIEW

**Comment by berestv** [comment]
@Hugofslopes @Zahhask45 everything in `/coverage_report ` is generated automatically and therefore doesn't need to be reviewed. Also, please DO NOT delete the branch after merging (code will be added by @Biltes)

File: tests/unit/speed_sensor/test/test_speed_sensor.c [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+    if (cmock_num_calls == 1)
+    {
+        longjmp(test_break_jump, 1);
+    }
+
+    // Otherwise (Index 0), just pretend we set the flags successfully
+    return TX_SUCCESS;
+}
+
+void tearDown(void)
```
Comment (melaniereis) [inline tests/unit/speed_sensor/test/test_speed_sensor.c]: what is this for?

Comment (berestv) [reply tests/unit/speed_sensor/test/test_speed_sensor.c]: Part of the best practices for writing Unity Tests. When Ceedling compiles my files, it automatically generates a .c file that coordinates the tests. This generated code explicitly calls tearDown() after every test case, so it should be defined even if not in use. 
In embedded testing, you rarely need code inside tearDown because setUp usually resets everything. You would only put code there if:
- You used malloc in a test and you need to free it..
- You opened a file handle and you need to fclose it.
- You set a global pointer that persists across tests and could crash the next test.
